core/src/main/scala/cats/ApplicativeError.scala

```diff
@@ -287,6 +287,30 @@ trait ApplicativeError[F[_], E] extends Applicative[F] {
   def catchOnly[T >: Null <: Throwable]: CatchOnlyPartiallyApplied[T, F, E] =
     new CatchOnlyPartiallyApplied[T, F, E](this)
 
+  /**
+   * Often E can be created from Throwable. Here we try to call pure or
+   * catch, adapt into E, and raise.
+   *
+   * Exceptions that cannot be adapted to E will be propagated outside of `F`
```

Comment:
Repeating my comments from https://github.com/typelevel/cats/issues/4286#issuecomment-1226096402:

This doesn't seem right to me. Exception-throwing code should ideally never be written outside of a `catchNonFatal`; pure code that uses this method (or any method in Cats) should not throw exceptions.

It seems to me what you actually want is maybe something like `F[G[A]]`, with `ApplicativeError[G, E]` and `ApplicativeThrow[F]`. That way non-adaptable exceptions can be safely handled in the `F` effect.

In which case I think you want something like this?
```scala
F.catchNonFatal(doTheThing()).map(G.pure).recover {
  case ex: MyAdaptableException => G.raiseError(whatever)
}
```

Maybe a dedicated method/syntax could make that a bit neater. I'm not entirely sure where it could live.

Reply:
Honestly, I personally have never used `catchOnly` before, and never paid attention to it. And now I'm a bit surprised that such a function even exists in here.

On the other hand though, there are several examples in Cats when a function can throw an exception, e.g. `NonEmptyList.fromListUnsafe` and alike. It is a slightly different case though, because happens not in a type class but rather in a particular data type. But still close.

From that point of view, if we want to have functions in Cats like `catchOnly` or similar that can (re)throw exceptions, then perhaps it makes sense to mark them "unsafe", because they really are, e.g.: `unsafeCatchOnly` or `catchOnlyUnsafe`. And also explain in the docs, why they are unsafe.

Reply:
Would the path forward then be to rename/deprecate the existing `catch*` methods?

Reply:
I think this PR would be better off staying focused on its initial goal.
Perhaps, it makes sense to provide a safe version of `catchNonFatalAs` as @armanbilge suggested along with its unsafe counterpart.

The effort on renaming of the existing methods could be addressed in a separate PR later on.

Reply:
I figured the easiest way to advance would be to implement `*Unsafe` versions of the proposed methods, and a `catchOnlySafe` version of `catchOnly`.

This way we have concrete implementations to look at and figure out which add value.